### PR TITLE
opt: remove TableMeta.IsSkipLocked

### PR DESCRIPTION
**Pull request by mgartner**
#### opt: fix duplicate join multiplicity test

Test case 9 in `TestGetJoinMultiplicity` was a duplicate of test case 2.
It has been updated to be similar, but use an `INNER JOIN` instead of a
`LEFT JOIN`, which I believe it was originally supposed to be - there is
no similar existing test case.

Release note: None

#### opt: remove TableMeta.IsSkipLocked

`TableMeta.IsSkipLocked` was used by the multiplicity builder to
determine whether a column was filtered or not by a `SKIP LOCKED` scan.
However, this field was unnecessary. The `deriveUnfilteredCols` function
in the multiplicity builder already traverses expressions all the way
down to `ScanExpr`s, only collecting unfiltered columns if they
originate from a scan where `ScanExpr.IsUnfiltered` returns true.
`ScanExpr.IsUnfiltered` returns false if the scan is a `SKIP LOCKED`
scan. Therefore, no additional mechanism is required to detect columns
filtered by `SKIP LOCKED` scans.

I noticed that `TableMeta.IsSkipLocked` was not needed because the
multiplicity builder unit tests added in #85720 never set it, yet the
tests passed.

This commit also adds more multiplicity builder unit tests for
self-joins with `SKIP LOCKED`.

Release note: None

#### opt: ensure min cardinality of zero for expressions with SKIP LOCKED

This commit ensures that the minimum cardinality of a scan or
lookup-join with `SKIP LOCKED` is zero. This shouldn't be needed because
the logical to derive the cardinality for these expressions should never
produce a non-zero minimum cardinality, but this provides extra safety.

Release note: None

#### opt: improve comments in multiplicity builder

Release note: None

### CODE REVIEW

**Review by rytaft — APPROVED** [review]
:lgtm: nice catch!

> Reviewed 1 of 1 files at r1, 5 of 5 files at r2, 1 of 1 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/85933#-:-N99zqSd6bIuswC_1S5s:b-pe67a)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @michae2 and @msirek)

**Review by michae2 — APPROVED** [review]
:lgtm: Thank you for this!

> Reviewed 1 of 1 files at r1, 5 of 5 files at r2, 1 of 1 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/85933#-:-N9AQS33241QC1JT_Ou2:bh1e47a)* status: :shipit: complete! 2 of 0 LGTMs obtained (waiting on @mgartner and @msirek)

___
*[`pkg/sql/opt/memo/multiplicity_builder.go` line 280 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/85933#-N9AOryC8PDg6fdFvLOu:-N9AOryC8PDg6fdFvLOv:bmhkwny) ([raw file](https://github.com/cockroachdb/cockroach/blob/8ad465c58ed7daeef3d179f42a25d6b17cc7098e/pkg/sql/opt/memo/multiplicity_builder.go#L280)):*
> ```Go
> 	}
> 	if checkSelfJoinCase(left.Memo().Metadata(), filters) {
> 		// Case 2a.
> ```

Is this self-join case handled? It looks like the other cases are, thanks to the cardinality checks and the use of unfiltered cols, but I don't see how this self-join case is handled.

But the cases you added to multiplicity_builder_test seem to cover this, so it must be handled somehow. :thinking:

**Review by msirek — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/85933)* status: :shipit: complete! 2 of 0 LGTMs obtained (waiting on @mgartner)

___
*[`pkg/sql/opt/memo/logical_props_builder.go` line 523 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/85933#-N9DrPql4YO3AeJ-auFU:-N9DrPqm4G_11Nj21Af0:bjt72zj) ([raw file](https://github.com/cockroachdb/cockroach/blob/37ccddeb7035b1e71d23acc9b3986f3954549622/pkg/sql/opt/memo/logical_props_builder.go#L523)):*
> ```Go
> 		// this provides extra safety.
> 		rel.Cardinality = rel.Cardinality.AsLowAs(0)
> 	}
> ```

Should similar code be added to `buildIndexJoinProps`?

**Review by mgartner — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/85933)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 2 stale) (waiting on @michae2, @msirek, and @rytaft)

___
*[`pkg/sql/opt/memo/logical_props_builder.go` line 523 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/85933#-N9DrPql4YO3AeJ-auFU:-N9Dv4o-1aNVQrUU4KWx:b58klqp) ([raw file](https://github.com/cockroachdb/cockroach/blob/7cf49aeeb27d7c95438eab1272c1c24bf8616b19/pkg/sql/opt/memo/logical_props_builder.go#L523)):*
<details><summary><i>Previously, msirek (Mark Sirek) wrote…</i></summary><blockquote>

Should similar code be added to `buildIndexJoinProps`?
</blockquote></details>

Yes! It was already added in #85720.
___
*[`pkg/sql/opt/memo/multiplicity_builder.go` line 280 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/85933#-N9AOryC8PDg6fdFvLOu:-N9DpvmV9giRpsXZrZNL:bmp5mwi) ([raw file](https://github.com/cockroachdb/cockroach/blob/8b406e7ab3c514a62c6da2156dd7a6e88674a7df/pkg/sql/opt/memo/multiplicity_builder.go#L280)):*
<details><summary><i>Previously, michae2 (Michael Erickson) wrote…</i></summary><blockquote>

Is this self-join case handled? It looks like the other cases are, thanks to the cardinality checks and the use of unfiltered cols, but I don't see how this self-join case is handled.

But the cases you added to multiplicity_builder_test seem to cover this, so it must be handled somehow. :thinking:
</blockquote></details>

This confused me too. IIRC it is handled by `verifyFiltersAreValidEqualities` just above here. If the columns in the equality are filtered on the right input, then we never actually make it here - we return false 3 lines up.

We check for valid equality filters before we check the self join (2a) or FK (2b) case, because both cases have the same requirements about the filters. This paragraph in the comment calls this out:

```
// In both the self-join and the foreign key cases, the left columns must be
// not-null, and the right columns must be either unfiltered, or the left and
// right must be Select expressions where the left side filters imply the right
// side filters and right columns are unfiltered in the right Select's input
// (see condition #3b in the comment for verifyFiltersAreValidEqualities).
```

I found this hard to follow (and I wrote it!), so I added a new commit that rewrites this in an attempt to be more clear.

**Review by michae2 — COMMENTED** [review]
> Reviewed all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/85933)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 2 stale) (waiting on @mgartner, @michae2, @msirek, and @rytaft)

___
*[`pkg/sql/opt/memo/multiplicity_builder.go` line 280 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/85933#-N9AOryC8PDg6fdFvLOu:-N9DxZilDO26O9Q3YP1U:bz0jrin) ([raw file](https://github.com/cockroachdb/cockroach/blob/8b406e7ab3c514a62c6da2156dd7a6e88674a7df/pkg/sql/opt/memo/multiplicity_builder.go#L280)):*
<details><summary><i>Previously, mgartner (Marcus Gartner) wrote…</i></summary><blockquote>

This confused me too. IIRC it is handled by `verifyFiltersAreValidEqualities` just above here. If the columns in the equality are filtered on the right input, then we never actually make it here - we return false 3 lines up.

We check for valid equality filters before we check the self join (2a) or FK (2b) case, because both cases have the same requirements about the filters. This paragraph in the comment calls this out:

```
// In both the self-join and the foreign key cases, the left columns must be
// not-null, and the right columns must be either unfiltered, or the left and
// right must be Select expressions where the left side filters imply the right
// side filters and right columns are unfiltered in the right Select's input
// (see condition #3b in the comment for verifyFiltersAreValidEqualities).
```

I found this hard to follow (and I wrote it!), so I added a new commit that rewrites this in an attempt to be more clear.
</blockquote></details>

I see, that makes sense. Thank you!

**Comment by msirek** [comment]
> Yes! It was already added in #85720.

Ah, thanks.. I didn't see it. I was reviewing code in a branch that's a few days old.

**Review by msirek — APPROVED** [review]
:lgtm: 

> Reviewed 1 of 5 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/85933#-:-N9Dy8KL21GSDy_BjwSQ:bg7igkv)* status: :shipit: complete! 1 of 0 LGTMs obtained (and 2 stale) (waiting on @mgartner, @michae2, @msirek, and @rytaft)

**Review by michae2 — APPROVED** [review]
> Reviewed 1 of 1 files at r4.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/85933)* status: :shipit: complete! 1 of 0 LGTMs obtained (and 2 stale) (waiting on @mgartner and @msirek)
